File: services/imageLookupService.js
const crypto = require("crypto");
const fs = require("fs");
const path = require("path");
const { BlobServiceClient } = require("@azure/storage-blob");
const sharp = require("sharp");

const GENERATED_IMAGE_DIR = path.join(__dirname, "..", "data", "generated-images");
const GENERATED_IMAGE_URL_PREFIX = "/generated-images";
const OPENVERSE_API_URL = "https://api.openverse.org/v1/images/";
const IMAGE_PAGE_SIZE = 16;
const OUTPUT_WIDTH = 640;
const OUTPUT_HEIGHT = 480;
const STORAGE_CONNECTION_STRING = process.env.AZURE_STORAGE_CONNECTION_STRING;
const STORAGE_CONTAINER = process.env.AZURE_STORAGE_CONTAINER || "generated-images";
const USE_BLOB_STORAGE = Boolean(STORAGE_CONNECTION_STRING);
const CATEGORY_QUERY_HINTS = {
  music: ["band", "musician", "logo", "album cover"],
  movies: ["film", "poster", "still"],
  sports: ["athlete", "team", "action"],
  gaming: ["game", "character", "art"]
};

const GENERIC_STOCK_TERMS = /(stock|vector|illustration|clipart|template|background)/;
let containerClientPromise;

async function getContainerClient() {
  if (!USE_BLOB_STORAGE) {
    return null;
  }

  if (!containerClientPromise) {
    containerClientPromise = (async () => {
      const blobServiceClient = BlobServiceClient.fromConnectionString(STORAGE_CONNECTION_STRING);
      const containerClient = blobServiceClient.getContainerClient(STORAGE_CONTAINER);
      await containerClient.createIfNotExists({ access: "blob" });
      return containerClient;
    })();
  }

  return containerClientPromise;
}

function buildQueries(entry) {
  const queries = new Set();
  const name = entry.name.trim();
  const category = entry.category.trim();
  const categoryKey = category.toLowerCase();
  const hints = CATEGORY_QUERY_HINTS[categoryKey] || [];

  if (name && category) {
    queries.add(`${name} ${category}`);
    queries.add(`"${name}" ${category}`);
  }

  for (const hint of hints) {
    queries.add(`${name} ${hint}`);
    queries.add(`"${name}" ${hint}`);
  }

  if (name) {
    queries.add(name);
  }

  return [...queries];
}

function tokenize(value) {
  return (value || "")
    .toLowerCase()
    .split(/[^a-z0-9]+/)
    .filter((token) => token.length > 1);
}

function normaliseTags(tags) {
  return (tags || []).map((tag) => {
    if (typeof tag === "string") {
      return tag;
    }

    return tag?.name || "";
  });
}

function getCandidateText(candidate) {
  return [candidate.title, candidate.creator, candidate.source, ...normaliseTags(candidate.tags)]
    .filter(Boolean)
    .join(" ")
    .toLowerCase();
}

function getCandidateScore(candidate, entry, query) {
  const haystack = getCandidateText(candidate);
  const namePhrase = entry.name.toLowerCase();
  const categoryPhrase = entry.category.toLowerCase();
  const queryTokens = tokenize(query);
  const nameTokens = tokenize(entry.name);
  const categoryTokens = tokenize(entry.category);
  let score = 0;
  let nameTokenMatches = 0;
  let categoryTokenMatches = 0;

  if (haystack.includes(namePhrase)) {
    score += 50;
  }

  if (categoryPhrase && haystack.includes(categoryPhrase)) {
    score += 16;
  }

  for (const token of queryTokens) {
    if (haystack.includes(token)) {
      score += 8;
    }
  }

  for (const token of nameTokens) {
    if (haystack.includes(token)) {
      score += 10;
      nameTokenMatches += 1;
    }
  }

  for (const token of categoryTokens) {
    if (haystack.includes(token)) {
      score += 4;
      categoryTokenMatches += 1;
    }
  }

  if (nameTokens.length > 0 && nameTokenMatches === 0) {
    score -= 40;
  }

  if (categoryTokens.length > 0 && categoryTokenMatches === 0) {
    score -= 8;
  }

  const width = Number(candidate.width || 0);
  const height = Number(candidate.height || 0);
  if (width >= 400 && height >= 300) {
    score += 18;
  }

  if (width && height) {
    const areaScore = Math.min((width * height) / 250000, 18);
    const ratio = width / height;
    const ratioScore = Math.max(0, 10 - Math.abs(ratio - 1.2) * 10);
    score += areaScore + ratioScore;
  }

  const title = (candidate.title || "").toLowerCase();
  if (/(stock|vector|illustration|clipart|template|background|icon)/.test(title)) {
    score -= 20;
  }

  if (GENERIC_STOCK_TERMS.test(haystack)) {
    score -= 18;
  }

  if (categoryPhrase === "music") {
    if (/(band|musician|artist|album|cover|logo)/.test(haystack)) {
      score += 16;
    }

    if (/(music background|musical background|abstract music|equalizer)/.test(haystack)) {
      score -= 20;
    }
  }

  if (!candidate.url && !candidate.thumbnail) {
    score -= 100;
  }

  return score;
}

async function searchOpenverse(entry) {
  const ranked = [];
  const seen = new Set();

  for (const query of buildQueries(entry)) {
    const response = await fetch(
      `${OPENVERSE_API_URL}?q=${encodeURIComponent(query)}&page_size=${IMAGE_PAGE_SIZE}`,
      {
        headers: {
          Accept: "application/json",
          "User-Agent": "listflair-app/1.0"
        }
      }
    );

    if (!response.ok) {
      continue;
    }

    const payload = await response.json();
    const results = payload.results || [];
    for (const candidate of results) {
      const dedupeKey = candidate.id || candidate.foreign_landing_url || candidate.url || candidate.thumbnail;
      if (!dedupeKey || seen.has(dedupeKey)) {
        continue;
      }

      seen.add(dedupeKey);
      ranked.push({
        query,
        score: getCandidateScore(candidate, entry, query),
        sourceUrl: candidate.foreign_landing_url || candidate.url || null,
        fetchUrl: candidate.url || candidate.thumbnail || null,
        thumbnailUrl: candidate.thumbnail || null,
        candidate
      });
    }
  }

  ranked.sort((left, right) => right.score - left.score);
  return ranked;
}

async function ensureGeneratedImageDir() {
  await fs.promises.mkdir(GENERATED_IMAGE_DIR, { recursive: true });
}

async function downloadBuffer(url) {
  const response = await fetch(url, {
    headers: {
      Accept: "image/*",
      "User-Agent": "listflair-app/1.0"
    }
  });

  if (!response.ok) {
    throw new Error(`Image download failed with status ${response.status}`);
  }

  const arrayBuffer = await response.arrayBuffer();
  return Buffer.from(arrayBuffer);
}

async function cacheImageLocally(entry, rankedCandidate) {
  let imageBuffer;
  try {
    imageBuffer = await downloadBuffer(rankedCandidate.fetchUrl);
  } catch (error) {
    if (!rankedCandidate.thumbnailUrl || rankedCandidate.thumbnailUrl === rankedCandidate.fetchUrl) {
      throw error;
    }
    imageBuffer = await downloadBuffer(rankedCandidate.thumbnailUrl);
  }

  const fileHash = crypto
    .createHash("sha1")
    .update(`${entry.id}:${rankedCandidate.fetchUrl}:${Date.now()}`)
    .digest("hex")
    .slice(0, 12);
  const fileName = `${entry.id}-${fileHash}.jpg`;
  const transformedBuffer = await sharp(imageBuffer)
    .rotate()
    .resize(OUTPUT_WIDTH, OUTPUT_HEIGHT, {
      fit: "inside",
      withoutEnlargement: true
    })
    .jpeg({ quality: 82, mozjpeg: true })
    .toBuffer();

  if (USE_BLOB_STORAGE) {
    const containerClient = await getContainerClient();
    const blobClient = containerClient.getBlockBlobClient(fileName);
    await blobClient.uploadData(transformedBuffer, {
      blobHTTPHeaders: { blobContentType: "image/jpeg" }
    });

    return {
      imageUrl: blobClient.url,
      imageSource: rankedCandidate.sourceUrl,
      imageQuery: rankedCandidate.query
    };
  }

  await ensureGeneratedImageDir();
  const filePath = path.join(GENERATED_IMAGE_DIR, fileName);
  await fs.promises.writeFile(filePath, transformedBuffer);

  return {
    imageUrl: `${GENERATED_IMAGE_URL_PREFIX}/${fileName}`,
    imageSource: rankedCandidate.sourceUrl,
    imageQuery: rankedCandidate.query
  };
}

async function findAndCacheImageForEntry(entry, resultIndex = 0) {
  const rankedCandidates = await searchOpenverse(entry);
  if (!rankedCandidates.length) {
    throw new Error("No suitable image found");
  }

  const safeIndex = Math.max(0, resultIndex);
  const rankedCandidate = rankedCandidates[safeIndex % rankedCandidates.length];
  const cachedImage = await cacheImageLocally(entry, rankedCandidate);

  return {
    ...cachedImage,
    imageResultIndex: safeIndex % rankedCandidates.length
  };
}

function isGeneratedImageUrl(imageUrl) {
  if (typeof imageUrl !== "string") {
    return false;
  }

  if (imageUrl.startsWith(`${GENERATED_IMAGE_URL_PREFIX}/`)) {
    return true;
  }

  if (USE_BLOB_STORAGE) {
    return imageUrl.includes(`/${STORAGE_CONTAINER}/`);
  }

  return false;
}

async function removeCachedImage(imageUrl) {
  if (!isGeneratedImageUrl(imageUrl)) {
    return;
  }

  if (USE_BLOB_STORAGE && imageUrl.startsWith("http")) {
    const url = new URL(imageUrl);
    const pathParts = url.pathname.split("/").filter(Boolean);
    const containerIndex = pathParts.findIndex((part) => part === STORAGE_CONTAINER);
    if (containerIndex >= 0 && pathParts.length > containerIndex + 1) {
      const blobName = pathParts.slice(containerIndex + 1).join("/");
      const containerClient = await getContainerClient();
      await containerClient.deleteBlob(blobName, { deleteSnapshots: "include" }).catch((error) => {
        if (error.statusCode !== 404) {
          throw error;
        }
      });
    }
    return;
  }

  const filePath = path.join(GENERATED_IMAGE_DIR, imageUrl.replace(`${GENERATED_IMAGE_URL_PREFIX}/`, ""));
  try {
    await fs.promises.unlink(filePath);
  } catch (error) {
    if (error.code !== "ENOENT") {
      throw error;
    }
  }
}

async function searchWebImages(query) {
  // DDG image search (unofficial) — no API key required
  const pageRes = await fetch(
    `https://duckduckgo.com/?q=${encodeURIComponent(query)}&iax=images&ia=images`,
    {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
        "Accept-Language": "en-US,en;q=0.9",
        Accept: "text/html,application/xhtml+xml"
      },
      redirect: "follow"
    }
  );

  if (!pageRes.ok) {
    throw new Error(`Image search unavailable (${pageRes.status})`);
  }

  const html = await pageRes.text();
  const vqdMatch =
    html.match(/vqd=['"]([^'"]+)['"]/) ||
    html.match(/"vqd":"([^"]+)"/) ||
    html.match(/vqd=([a-zA-Z0-9_-]+)/);
  if (!vqdMatch) {
    throw new Error("Image search temporarily unavailable");
  }
  const vqd = vqdMatch[1];

  const imagesUrl = new URL("https://duckduckgo.com/i.js");
  imagesUrl.searchParams.set("q", query);
  imagesUrl.searchParams.set("vqd", vqd);
  imagesUrl.searchParams.set("p", "1");
  imagesUrl.searchParams.set("s", "0");
  imagesUrl.searchParams.set("l", "us-en");
  imagesUrl.searchParams.set("f", ",,,,,");

  const imgRes = await fetch(imagesUrl.toString(), {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
      Referer: "https://duckduckgo.com/",
      Accept: "application/json, text/javascript, */*; q=0.01",
      "Accept-Language": "en-US,en;q=0.9"
    }
  });

  if (!imgRes.ok) {
    throw new Error(`Image search failed (${imgRes.status})`);
  }

  const data = await imgRes.json();
  return (data.results || [])
    .filter((r) => r.thumbnail || r.image)
    .slice(0, 24)
    .map((r) => ({
      title: r.title || "",
      thumbnailUrl: r.thumbnail || r.image || "",
      fetchUrl: r.image || r.thumbnail || "",
      sourceUrl: r.url || null
    }));
}

async function cacheSelectedImage(entry, { fetchUrl, thumbnailUrl, sourceUrl, query }) {
  return cacheImageLocally(entry, {
    fetchUrl,
    thumbnailUrl: thumbnailUrl || fetchUrl,
    sourceUrl: sourceUrl || null,
    query: query || ""
  });
}

module.exports = {
  findAndCacheImageForEntry,
  searchWebImages,
  cacheSelectedImage,
  isGeneratedImageUrl,
  removeCachedImage,
  GENERATED_IMAGE_DIR,
  GENERATED_IMAGE_URL_PREFIX
};

File: api/Services/ImageLookupService.cs
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;
using Azure.Storage.Blobs;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Formats.Jpeg;
using SixLabors.ImageSharp.Processing;
using ListFlair.Api.Models;

namespace ListFlair.Api.Services;

public class ImageLookupService : IImageLookupService
{
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly IConfiguration _config;
    private readonly bool _useBlobStorage;
    private readonly string? _storageConnectionString;
    private readonly string _storageContainer;
    private BlobContainerClient? _containerClient;
    private readonly SemaphoreSlim _containerLock = new(1, 1);

    private const string OpenverseApiUrl = "https://api.openverse.org/v1/images/";
    private const int ImagePageSize = 16;
    private const int OutputWidth = 640;
    private const int OutputHeight = 480;
    private const string GeneratedImageUrlPrefix = "/generated-images";

    private static readonly Dictionary<string, string[]> CategoryQueryHints = new()
    {
        ["film"] = ["movie", "poster", "still", "cinema"],
        ["movie"] = ["film", "poster", "still", "cinema"],
        ["music"] = ["band", "musician", "logo", "album cover"],
        ["movies"] = ["film", "poster", "still"],
        ["sports"] = ["athlete", "team", "action"],
        ["gaming"] = ["game", "character", "art"]
    };

    private static readonly Dictionary<string, string> TagAliases = new()
    {
        ["films"] = "film",
        ["movie"] = "film",
        ["movies"] = "film",
        ["cinema"] = "film",
        ["motionpicture"] = "film",
        ["motionpictures"] = "film"
    };

    public string GeneratedImageDir { get; }

    public ImageLookupService(IHttpClientFactory httpClientFactory, IConfiguration config)
    {
        _httpClientFactory = httpClientFactory;
        _config = config;
        _storageConnectionString = config["AZURE_STORAGE_CONNECTION_STRING"]
            ?? Environment.GetEnvironmentVariable("AZURE_STORAGE_CONNECTION_STRING");
        _useBlobStorage = !string.IsNullOrEmpty(_storageConnectionString);
        _storageContainer = config["StorageContainer"]
            ?? Environment.GetEnvironmentVariable("AZURE_STORAGE_CONTAINER")
            ?? "generated-images";

        var repoRoot = config["RepoRoot"]
            ?? Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", ".."));
        GeneratedImageDir = Path.Combine(repoRoot, "data", "generated-images");
    }

    // ──────────────────────────────────────────────────────────────
    // Tag parsing / query building
    // ──────────────────────────────────────────────────────────────

    private static List<string> ParseTags(string? value)
    {
        return (value ?? "")
            .Split([';', ',', '|', '/'], StringSplitOptions.RemoveEmptyEntries)
            .Select(t => t.Trim())
            .Where(t => !string.IsNullOrEmpty(t))
            .Take(8)
            .ToList();
    }

    private static string NormaliseTag(string tag)
    {
        var token = new string(tag.ToLowerInvariant().Where(char.IsLetterOrDigit).ToArray());
        return TagAliases.TryGetValue(token, out var alias) ? alias : token;
    }

    private static List<string> BuildQueries(Entry entry)
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var queries = new List<string>();

        void Add(string q) { if (seen.Add(q)) queries.Add(q); }

        var name = entry.Name.Trim();
        var tags = ParseTags(entry.Category);
        var tagsPhrase = string.Join(" ", tags).Trim();
        var canonicalTags = tags.Select(NormaliseTag).Where(t => !string.IsNullOrEmpty(t)).Distinct().ToList();
        var hintTags = canonicalTags.Count > 0 ? canonicalTags : [NormaliseTag(entry.Category.Trim())];
        var hints = hintTags.SelectMany(t => CategoryQueryHints.TryGetValue(t, out var h) ? h : []).ToList();

        if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(tagsPhrase))
        {
            Add($"{name} {tagsPhrase}");
            Add($"\"{name}\" {tagsPhrase}");
        }

        foreach (var tag in tags)
        {
            Add($"{name} {tag}");
            Add($"\"{name}\" {tag}");
        }

        foreach (var hint in hints)
        {
            Add($"{name} {hint}");
            Add($"\"{name}\" {hint}");
        }

        if (!string.IsNullOrEmpty(name)) Add(name);

        return queries;
    }

    // ──────────────────────────────────────────────────────────────
    // Candidate scoring
    // ──────────────────────────────────────────────────────────────

    private static List<string> Tokenize(string? value)
    {
        return (value ?? "")
            .ToLowerInvariant()
            .Split(c => !char.IsLetterOrDigit(c), StringSplitOptions.RemoveEmptyEntries)
            .Where(t => t.Length > 1)
            .ToList();
    }

    private static string GetCandidateText(OpenverseCandidate c)
    {
        var parts = new List<string?> { c.Title, c.Creator, c.Source };
        parts.AddRange(c.Tags?.Select(t => t.Name) ?? []);
        return string.Join(" ", parts.Where(p => !string.IsNullOrEmpty(p))).ToLowerInvariant();
    }

    private static int GetCandidateScore(OpenverseCandidate candidate, Entry entry, string query)
    {
        var haystack = GetCandidateText(candidate);
        var namePhrase = entry.Name.ToLowerInvariant();
        var categoryPhrase = entry.Category.ToLowerInvariant();
        var queryTokens = Tokenize(query);
        var nameTokens = Tokenize(entry.Name);
        var categoryTokens = Tokenize(entry.Category);
        var score = 0;

        if (haystack.Contains(namePhrase)) score += 50;
        if (!string.IsNullOrEmpty(categoryPhrase) && haystack.Contains(categoryPhrase)) score += 16;

        foreach (var t in queryTokens) if (haystack.Contains(t)) score += 8;

        var nameHits = nameTokens.Count(t => haystack.Contains(t));
        score += nameHits * 10;
        var catHits = categoryTokens.Count(t => haystack.Contains(t));
        score += catHits * 4;

        if (nameTokens.Count > 0 && nameHits == 0) score -= 40;
        if (categoryTokens.Count > 0 && catHits == 0) score -= 8;

        var width = candidate.Width;
        var height = candidate.Height;
        if (width >= 400 && height >= 300) score += 18;
        if (width > 0 && height > 0)
        {
            var areaScore = (int)Math.Min((double)width * height / 250000.0, 18.0);
            var ratio = (double)width / height;
            var ratioScore = (int)Math.Max(0.0, 10.0 - Math.Abs(ratio - 1.2) * 10.0);
            score += areaScore + ratioScore;
        }

        var title = (candidate.Title ?? "").ToLowerInvariant();
        if (System.Text.RegularExpressions.Regex.IsMatch(title, @"(stock|vector|illustration|clipart|template|background|icon)"))
            score -= 20;
        if (System.Text.RegularExpressions.Regex.IsMatch(haystack, @"(stock|vector|illustration|clipart|template|background)"))
            score -= 18;

        if (categoryPhrase == "music")
        {
            if (System.Text.RegularExpressions.Regex.IsMatch(haystack, @"(band|musician|artist|album|cover|logo)"))
                score += 16;
            if (System.Text.RegularExpressions.Regex.IsMatch(haystack, @"(music background|musical background|abstract music|equalizer)"))
                score -= 20;
        }

        if (string.IsNullOrEmpty(candidate.Url) && string.IsNullOrEmpty(candidate.Thumbnail))
            score -= 100;

        return score;
    }

    // ──────────────────────────────────────────────────────────────
    // Openverse search
    // ──────────────────────────────────────────────────────────────

    private async Task<List<RankedCandidate>> SearchOpenverseAsync(Entry entry)
    {
        var ranked = new List<RankedCandidate>();
        var seen = new HashSet<string>();
        var client = _httpClientFactory.CreateClient("openverse");

        foreach (var query in BuildQueries(entry))
        {
            HttpResponseMessage response;
            try
            {
                response = await client.GetAsync(
                    $"{OpenverseApiUrl}?q={Uri.EscapeDataString(query)}&page_size={ImagePageSize}");
            }
            catch { continue; }

            if (!response.IsSuccessStatusCode) continue;

            var json = await response.Content.ReadAsStringAsync();
            var payload = JsonSerializer.Deserialize<OpenverseResponse>(json, JsonOpts);
            if (payload?.Results == null) continue;

            foreach (var candidate in payload.Results)
            {
                var dedupeKey = candidate.Id ?? candidate.ForeignLandingUrl ?? candidate.Url ?? candidate.Thumbnail;
                if (string.IsNullOrEmpty(dedupeKey) || !seen.Add(dedupeKey)) continue;

                ranked.Add(new RankedCandidate
                {
                    Query = query,
                    Score = GetCandidateScore(candidate, entry, query),
                    FetchUrl = candidate.Url ?? candidate.Thumbnail ?? "",
                    ThumbnailUrl = candidate.Thumbnail ?? candidate.Url,
                    SourceUrl = candidate.ForeignLandingUrl ?? candidate.Url
                });
            }
        }

        ranked.Sort((a, b) => b.Score.CompareTo(a.Score));
        return ranked;
    }

    // ──────────────────────────────────────────────────────────────
    // Image caching
    // ──────────────────────────────────────────────────────────────

    private async Task<BlobContainerClient?> GetContainerClientAsync()
    {
        if (!_useBlobStorage) return null;
        if (_containerClient != null) return _containerClient;

        await _containerLock.WaitAsync();
        try
        {
            if (_containerClient != null) return _containerClient;
            var blobService = new BlobServiceClient(_storageConnectionString);
            _containerClient = blobService.GetBlobContainerClient(_storageContainer);
            await _containerClient.CreateIfNotExistsAsync(Azure.Storage.Blobs.Models.PublicAccessType.Blob);
            return _containerClient;
        }
        finally { _containerLock.Release(); }
    }

    private async Task<byte[]> DownloadBufferAsync(string url)
    {
        var client = _httpClientFactory.CreateClient("imagedownload");
        var response = await client.GetAsync(url);
        if (!response.IsSuccessStatusCode)
            throw new InvalidOperationException($"Image download failed with status {(int)response.StatusCode}");
        return await response.Content.ReadAsByteArrayAsync();
    }

    private async Task<byte[]> TransformImageAsync(byte[] input)
    {
        using var image = Image.Load(input);
        image.Mutate(ctx => ctx
            .AutoOrient()
            .Resize(new ResizeOptions
            {
                Mode = ResizeMode.Max,
                Size = new Size(OutputWidth, OutputHeight),
                Sampler = SixLabors.ImageSharp.Processing.Processors.Transforms.LanczosResampler.Lanczos3
            }));

        using var ms = new MemoryStream();
        await image.SaveAsync(ms, new JpegEncoder { Quality = 82 });
        return ms.ToArray();
    }

    private async Task<CachedImage> CacheImageLocallyAsync(Entry entry, RankedCandidate candidate)
    {
        byte[] imageBuffer;
        try { imageBuffer = await DownloadBufferAsync(candidate.FetchUrl); }
        catch
        {
            if (string.IsNullOrEmpty(candidate.ThumbnailUrl) || candidate.ThumbnailUrl == candidate.FetchUrl)
                throw;
            imageBuffer = await DownloadBufferAsync(candidate.ThumbnailUrl);
        }

        var hash = Convert.ToHexString(SHA1.HashData(
            System.Text.Encoding.UTF8.GetBytes($"{entry.Id}:{candidate.FetchUrl}:{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}")))
            .ToLowerInvariant()[..12];
        var fileName = $"{entry.Id}-{hash}.jpg";
        var transformed = await TransformImageAsync(imageBuffer);

        if (_useBlobStorage)
        {
            var container = await GetContainerClientAsync();
            var blob = container!.GetBlobClient(fileName);
            using var ms = new MemoryStream(transformed);
            await blob.UploadAsync(ms, new Azure.Storage.Blobs.Models.BlobUploadOptions
            {
                HttpHeaders = new Azure.Storage.Blobs.Models.BlobHttpHeaders { ContentType = "image/jpeg" }
            });
            return new CachedImage { ImageUrl = blob.Uri.ToString(), ImageSource = candidate.SourceUrl, ImageQuery = candidate.Query };
        }

        Directory.CreateDirectory(GeneratedImageDir);
        var filePath = Path.Combine(GeneratedImageDir, fileName);
        await File.WriteAllBytesAsync(filePath, transformed);
        return new CachedImage
        {
            ImageUrl = $"{GeneratedImageUrlPrefix}/{fileName}",
            ImageSource = candidate.SourceUrl,
            ImageQuery = candidate.Query
        };
    }

    // ──────────────────────────────────────────────────────────────
    // Public interface
    // ──────────────────────────────────────────────────────────────

    public async Task<CachedImage> FindAndCacheImageForEntryAsync(Entry entry, int requestedIndex = 0)
    {
        var candidates = await SearchOpenverseAsync(entry);
        if (candidates.Count == 0)
            throw new InvalidOperationException("No suitable image found");

        var safeIndex = Math.Max(0, requestedIndex);
        var candidate = candidates[safeIndex % candidates.Count];
        var cached = await CacheImageLocallyAsync(entry, candidate);
        cached.ImageResultIndex = safeIndex % candidates.Count;
        return cached;
    }

    public async Task<CachedImage> CacheSelectedImageAsync(Entry entry, string fetchUrl,
        string? thumbnailUrl, string? sourceUrl, string? query)
    {
        return await CacheImageLocallyAsync(entry, new RankedCandidate
        {
            FetchUrl = fetchUrl,
            ThumbnailUrl = thumbnailUrl ?? fetchUrl,
            SourceUrl = sourceUrl,
            Query = query ?? ""
        });
    }

    public async Task<List<WebImageResult>> SearchWebImagesAsync(string query)
    {
        var client = _httpClientFactory.CreateClient("duckduckgo");

        // Step 1: Get vqd token
        var pageResp = await client.GetAsync(
            $"https://duckduckgo.com/?q={Uri.EscapeDataString(query)}&iax=images&ia=images");
        if (!pageResp.IsSuccessStatusCode)
            throw new InvalidOperationException($"Image search unavailable ({(int)pageResp.StatusCode})");

        var html = await pageResp.Content.ReadAsStringAsync();
        var vqdMatch = System.Text.RegularExpressions.Regex.Match(html, @"vqd=['""]([^'""]+)['""]");
        if (!vqdMatch.Success)
            vqdMatch = System.Text.RegularExpressions.Regex.Match(html, @"""vqd"":""([^""]+)""");
        if (!vqdMatch.Success)
            vqdMatch = System.Text.RegularExpressions.Regex.Match(html, @"vqd=([a-zA-Z0-9_-]+)");
        if (!vqdMatch.Success)
            throw new InvalidOperationException("Image search temporarily unavailable");

        var vqd = vqdMatch.Groups[1].Value;

        // Step 2: Fetch images
        var imagesUrl = new UriBuilder("https://duckduckgo.com/i.js");
        imagesUrl.Query = $"q={Uri.EscapeDataString(query)}&vqd={Uri.EscapeDataString(vqd)}&p=1&s=0&l=us-en&f=,,,,,";
        var imgResp = await client.GetAsync(imagesUrl.Uri);
        if (!imgResp.IsSuccessStatusCode)
            throw new InvalidOperationException($"Image search failed ({(int)imgResp.StatusCode})");

        var json = await imgResp.Content.ReadAsStringAsync();
        var data = JsonSerializer.Deserialize<DdgImageResponse>(json, JsonOpts);

        return (data?.Results ?? [])
            .Select(r => new WebImageResult
            {
                Title = r.Title ?? "",
                ThumbnailUrl = NormaliseResultUrl(r.Thumbnail ?? r.Image),
                FetchUrl = NormaliseResultUrl(r.Image ?? r.Thumbnail),
                SourceUrl = NormaliseResultUrl(r.Url)
            })
            .Where(r => !string.IsNullOrEmpty(r.ThumbnailUrl) || !string.IsNullOrEmpty(r.FetchUrl))
            .Take(24)
            .ToList();
    }

    private static string? NormaliseResultUrl(string? url)
    {
        if (string.IsNullOrWhiteSpace(url)) return null;
        var t = url.Trim();
        if (t.StartsWith("//")) return "https:" + t;
        if (t.StartsWith("http://")) return "https://" + t["http://".Length..];
        if (t.StartsWith("https://")) return t;
        return null;
    }

    public bool IsGeneratedImageUrl(string? imageUrl)
    {
        if (string.IsNullOrEmpty(imageUrl)) return false;
        if (imageUrl.StartsWith($"{GeneratedImageUrlPrefix}/")) return true;
        if (_useBlobStorage) return imageUrl.Contains($"/{_storageContainer}/");
        return false;
    }

    public async Task RemoveCachedImageAsync(string? imageUrl)
    {
        if (!IsGeneratedImageUrl(imageUrl)) return;

        if (_useBlobStorage && imageUrl!.StartsWith("http"))
        {
            try
            {
                var uri = new Uri(imageUrl);
                var pathParts = uri.AbsolutePath.Split('/', StringSplitOptions.RemoveEmptyEntries);
                var containerIndex = Array.IndexOf(pathParts, _storageContainer);
                if (containerIndex >= 0 && pathParts.Length > containerIndex + 1)
                {
                    var blobName = string.Join("/", pathParts[(containerIndex + 1)..]);
                    var container = await GetContainerClientAsync();
                    await container!.DeleteBlobIfExistsAsync(blobName,
                        Azure.Storage.Blobs.Models.DeleteSnapshotsOption.IncludeSnapshots);
                }
            }
            catch { /* best effort */ }
            return;
        }

        var fileName = imageUrl!.Replace($"{GeneratedImageUrlPrefix}/", "");
        var filePath = Path.Combine(GeneratedImageDir, fileName);
        try { File.Delete(filePath); }
        catch (FileNotFoundException) { }
    }

    // ──────────────────────────────────────────────────────────────
    // Internal types
    // ──────────────────────────────────────────────────────────────

    private sealed class RankedCandidate
    {
        public string Query { get; set; } = "";
        public int Score { get; set; }
        public string FetchUrl { get; set; } = "";
        public string? ThumbnailUrl { get; set; }
        public string? SourceUrl { get; set; }
    }

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    private sealed class OpenverseResponse
    {
        public List<OpenverseCandidate>? Results { get; set; }
    }

    private sealed class DdgImageResponse
    {
        public List<DdgImageResult>? Results { get; set; }
    }

    private sealed class DdgImageResult
    {
        public string? Image { get; set; }
        public string? Thumbnail { get; set; }
        public string? Url { get; set; }
        public string? Title { get; set; }
    }
}

// Extension for string.Split with char predicate (used in Tokenize)
internal static class StringExtensions
{
    public static string[] Split(this string source, Func<char, bool> predicate,
        StringSplitOptions options = StringSplitOptions.None)
    {
        var parts = new List<string>();
        var start = 0;
        for (var i = 0; i <= source.Length; i++)
        {
            if (i == source.Length || predicate(source[i]))
            {
                if (i > start || options != StringSplitOptions.RemoveEmptyEntries)
                    parts.Add(source[start..i]);
                start = i + 1;
            }
        }
        return parts.ToArray();
    }
}
